Allow variable length signing

jonasnick opened the pull request:
~Draft because~
- ~we should add a test vector for partial sig verification once #9 is merged~
- ~perhaps add a sentence to "Signing Flow" section clarifying that you need to sign 32 byte messages for default BIP-340 verification (not sure right now how variable-length messages will impact existing BIP-340 verification)~

robot-dreams commented on bip-musig2.mediawiki:
```diff
@@ -276,15 +276,15 @@ Input:
 * The secret signing key ''sk'': a 32-byte array or 0-byte array (optional argument)
 * The aggregate public key ''aggpk'': a 32-byte array or 0-byte array (optional argument)
-* The message ''m'': a 32-byte array or 0-byte array (optional argument)
+* The message ''m'': a byte array with ''0 &le; len(extra_in) &le; 2<sup>32</sup>-1'' (optional argument)
```
Why does this condition (and the ones below) refer to `extra_in` instead of `m`?

jonasnick replied:
Just a copy and paste mistake, fixed

robot-dreams commented on bip-musig2.mediawiki:
```diff
@@ -276,15 +276,15 @@ Input:
 ** Let ''rand'' be the byte-wise xor of ''sk'' and ''hash<sub>MuSig/aux</sub>(rand')''<ref>The random data is hashed (with a unique tag) as a precaution against situations where the randomness may be correlated with the secret signing key itself. It is xored with the secret key (rather than combined with it in a hash) to reduce the number of operations exposed to the actual secret key.</ref>.
 * Else: let ''rand = rand' ''
-* Let ''k<sub>i</sub> = int(hash<sub>MuSig/nonce</sub>(rand || bytes(1, len(aggpk)) || aggpk || bytes(1, len(m)) || m || bytes(4, len(extra_in)) || extra_in || bytes(1, i - 1))) mod n'' for ''i = 1,2''
+* Let ''k<sub>i</sub> = int(hash<sub>MuSig/nonce</sub>(rand || bytes(1, len(aggpk)) || aggpk || bytes(4, len(extra_in)) || extra_in || m || bytes(1, i - 1))) mod n'' for ''i = 1,2''
```
What's the motivation for removing the length prefix (I want to make sure the intended benefit outweights the potential downside of forgetting the length for this one case)?

robot-dreams commented on bip-musig2.mediawiki:
```diff
@@ -276,15 +276,15 @@ Input:
 ** Let ''rand'' be the byte-wise xor of ''sk'' and ''hash<sub>MuSig/aux</sub>(rand')''<ref>The random data is hashed (with a unique tag) as a precaution against situations where the randomness may be correlated with the secret signing key itself. It is xored with the secret key (rather than combined with it in a hash) to reduce the number of operations exposed to the actual secret key.</ref>.
 * Else: let ''rand = rand' ''
-* Let ''k<sub>i</sub> = int(hash<sub>MuSig/nonce</sub>(rand || bytes(1, len(aggpk)) || aggpk || bytes(1, len(m)) || m || bytes(4, len(extra_in)) || extra_in || bytes(1, i - 1))) mod n'' for ''i = 1,2''
+* Let ''k<sub>i</sub> = int(hash<sub>MuSig/nonce</sub>(rand || bytes(1, len(aggpk)) || aggpk || bytes(4, len(extra_in)) || extra_in || m || bytes(1, i - 1))) mod n'' for ''i = 1,2''
```
What's the motivation for removing the length prefix (I want to make sure the intended benefit outweights the potential downside of forgetting the length)? Is it to let people hash a message even when they don't know the length ahead of time?

jonasnick replied:
No, the idea is that omitting the length prefix for a single argument is secure (since the total length of the hash function input is an implicit input and therefore no two different lists of arguments produce the same hash input). And it's a bit simpler. I don't know of a strong argument to prefer either option. However, I think consistency is good too, so I put back the length prefix.

robot-dreams commented on bip-musig2/reference.py:
```diff
@@ -227,8 +226,6 @@ def nonce_gen(sk: bytes, aggpk: bytes, msg: bytes, extra_in: bytes) -> Tuple[byt
     if len(aggpk) not in (0, 32):
         raise ValueError('The optional byte array aggpk must have length 0 or 32.')
-    if len(msg) not in (0, 32):
-        raise ValueError('The optional byte array msg must have length 0 or 32.')
```
One easy way to test longer messages might be to change `test_sign_and_verify_random`; the line that generates messages currently does `msg = secrets.token_bytes(32)` but any length should be OK. (Note that we'd also need to update `schnorr_verify` to remove the message length check.)

robot-dreams commented on bip-musig2/reference.py:
```diff
@@ -227,8 +226,6 @@ def nonce_gen(sk: bytes, aggpk: bytes, msg: bytes, extra_in: bytes) -> Tuple[byt
     if len(aggpk) not in (0, 32):
         raise ValueError('The optional byte array aggpk must have length 0 or 32.')
-    if len(msg) not in (0, 32):
-        raise ValueError('The optional byte array msg must have length 0 or 32.')
```
One easy way to test longer messages might be to change `test_sign_and_verify_random`; the line that generates messages currently does `msg = secrets.token_bytes(32)` but any length should be OK. (Note that we'd also need to update `schnorr_verify` to remove the message length check.)

jonasnick replied:
Interesting idea, but I'm not sure if we want to change schnorr_verify at this point.

robot-dreams commented on bip-musig2.mediawiki:
```diff
@@ -558,6 +558,7 @@ A change in the <code>MAJOR</code> version indicates that the specification is i
 The <code>MINOR</code> version is incremented whenever the inputs or the output of an algorithm changes in a backward-compatible way or new backward-compatible functionality is added.
 The <code>PATCH</code> version is incremented for other changes that are noteworthy (bug fixes, test vectors, important clarifications, etc.).
 
+* '''0.4.0''' (2022-05-25): Allow variable length messages, change order of arguments in ''NonceGen'' hash function and add test vector
```
If we go with that, is it worth also saying something like "removed length prefix for messages in NonceGen"?

robot-dreams commented on bip-musig2.mediawiki:
```diff
@@ -558,6 +558,7 @@ A change in the <code>MAJOR</code> version indicates that the specification is i
 The <code>MINOR</code> version is incremented whenever the inputs or the output of an algorithm changes in a backward-compatible way or new backward-compatible functionality is added.
 The <code>PATCH</code> version is incremented for other changes that are noteworthy (bug fixes, test vectors, important clarifications, etc.).
 
+* '''0.4.0''' (2022-05-25): Allow variable length messages, change order of arguments in ''NonceGen'' hash function and add test vector
```
If we go with that, is it worth also saying something like "removed length prefix for messages in NonceGen"?

jonasnick replied:
Added mention of message prefix change

robot-dreams reviewed (commented):
Seems reasonable overall (though might make sense to merge the BIP-340 change first).

Just confirming, the motivation for this change (as well as the corresponding change to BIP-340) is to give users a simpler API by letting them avoid having to hash their messages before signing?

jonasnick commented:
Rebased on master and removed draft status.

real-or-random commented on bip-musig2.mediawiki:
```diff
@@ -284,15 +284,15 @@ Input:
 ** Let ''rand'' be the byte-wise xor of ''sk'' and ''hash<sub>MuSig/aux</sub>(rand')''<ref>The random data is hashed (with a unique tag) as a precaution against situations where the randomness may be correlated with the secret signing key itself. It is xored with the secret key (rather than combined with it in a hash) to reduce the number of operations exposed to the actual secret key.</ref>.
 * Else: let ''rand = rand' ''
-* Let ''k<sub>i</sub> = int(hash<sub>MuSig/nonce</sub>(rand || bytes(1, len(aggpk)) || aggpk || bytes(1, len(m)) || m || bytes(4, len(extra_in)) || extra_in || bytes(1, i - 1))) mod n'' for ''i = 1,2''
+* Let ''k<sub>i</sub> = int(hash<sub>MuSig/nonce</sub>(rand || bytes(1, len(aggpk)) || aggpk || bytes(4, len(m) || m || bytes(4, len(extra_in)) || extra_in || bytes(1, i - 1))) mod n'' for ''i = 1,2''
```
missing right bracket
```suggestion
* Let ''k<sub>i</sub> = int(hash<sub>MuSig/nonce</sub>(rand || bytes(1, len(aggpk)) || aggpk || bytes(4, len(m)) || m || bytes(4, len(extra_in)) || extra_in || bytes(1, i - 1))) mod n'' for ''i = 1,2''
```

jonasnick replied:
Fixed

real-or-random commented:
Concept ACK

I'd prefer a 64 bit length. (secp256k1 currently takes a `size_t` as length argument.) This is sufficient for all users and hashing 4 bytes more won't hurt. 

---

> Seems reasonable overall (though might make sense to merge the BIP-340 change first).

I'll try to create PR this week, this is really overdue, sorry...

----

needs rebase

jonasnick commented:
Rebased and switched to 8 byte message length which seems excessive, but both options are fine with me.

jonasnick commented:
Rebased

real-or-random commented:
> I'll try to create PR this week, this is really overdue, sorry...

see https://github.com/sipa/bips/pull/221

real-or-random commented on bip-musig2.mediawiki:
```diff
@@ -315,15 +315,15 @@ Input:
 Input:
 * The secret signing key ''sk'': a 32-byte array or 0-byte array (optional argument)
 * The aggregate public key ''aggpk'': a 32-byte array or 0-byte array (optional argument)
-* The message ''m'': a 32-byte array or 0-byte array (optional argument)
+* The message ''m'': a byte array with ''0 &le; len(m) &le; 2<sup>32</sup>-1'' (optional argument)
```
Now the arguments are somewhat inconsistent here. For `sk` and `aggpk`, the "0-byte array" corresponds to the argument not being present (a magic value in a sense). Whereas `m` can be present but the 0-byte array. For the concrete algorithm here, it doesn't really make a difference but it would be good to have a clear interface. 

Maybe just drop the "or 0-byte array" for `sk` and `aggpk` and handle the presence of the arguments explicitly in the pseudocode. This seems clearer anyway. A real implementation will probably also use an `Option`, or `NULL`/`nil` or whatever special thing the language offers here.

jonasnick replied:
Pushed a WIP commit. TODO: need to define what it means for an argument to be undefined (which looks quite terrible, we should rather "define" an optional type) and define the empty array.  Is this what you mean? Seems quite complex in our pseudocode language.

I didn't think that the interface was particularly unclear. If the message is present and 0 bytes, then provide it to the nonce function as such. If it's not present, provide a 0-byte array just like the other optional arguments.

real-or-random replied:
> Pushed a WIP commit. TODO: need to define what it means for an argument to be undefined (which looks quite terrible, we should rather "define" an optional type) and define the empty array. Is this what you mean? Seems quite complex in our pseudocode language.

Yep, this is what I meant, and I think it's okay as is. Suggestion: Just say in prose "If the optional argument `sk` is present" (or "defined" -- though I'd slightly prefer "present"). This makes the connection to the "optional argument" wording from the input list, and so it's certainly clear enough if you ask me. 

We could alternatively define an optional type but either your current version or my suggestion is good enough, so I think having an optional type is overkill for pseudocode (that's exactly a detail of concrete programming language that we're trying to abstract away from.) 

We should however use something idiomatic in the Python code, i.e., [`None` and `Optional`](https://stackoverflow.com/questions/51710037/how-should-i-use-the-optional-type-hint).   

> 
> I didn't think that the interface was particularly unclear. If the message is present and 0 bytes, then provide it to the nonce function as such. If it's not present, provide a 0-byte array just like the other optional arguments.



> I didn't think that the interface was particularly unclear. If the message is present and 0 bytes, then provide it to the nonce function as such. If it's not present, provide a 0-byte array just like the other optional arguments.

I think the algorithm is indeed clear but I could imagine myself writing a Python implementation and wondering whether I should accept `None` or reject it.

jonasnick replied:
Ok, I cleaned up that commit.

real-or-random replied:
Looks really clean now!

jonasnick commented:
I added a commit that makes `NonceGen` distinguish between the cases where the message is not present and where it's present but has length 0. Also bumped the date on the other commits and updated the changelog accordingly.

real-or-random commented on bip-musig2.mediawiki:
```diff
@@ -328,9 +328,12 @@ Input:
 ** Let ''m = empty_bytes''
+** Let ''m_present = bytes(1,1)''
+* Else:
+** Let ''m_present = bytes(1,0)''
```
Shouldn't it be the other way around? (1 encoding true and 0 encoding false)? I mean it doesn't matter but it's really counter intuitive now. :D 

(Code formatting nit: could add a space after the `,`)

jonasnick replied:
Yes it should be. Sorry! Fixed.

real-or-random commented on bip-musig2.mediawiki:
```diff
@@ -313,16 +314,22 @@ Input:
 ** Let ''rand'' be the byte-wise xor of ''sk'' and ''hash<sub>MuSig/aux</sub>(rand')''<ref>The random data is hashed (with a unique tag) as a precaution against situations where the randomness may be correlated with the secret signing key itself. It is xored with the secret key (rather than combined with it in a hash) to reduce the number of operations exposed to the actual secret key.</ref>.
 * Else: let ''rand = rand' ''
+* If the optional argument ''aggpk'' is not present:
+** Let ''aggpk = empty_bytes''
```
nit: Is it easier to write "Let ''aggpk'' be the empty byte string"? This wouldn't need the `empty_bytestring`.  (Either way is fine, just not sure if this option came to your mind).

jonasnick replied:
I thought it's (very slightly) better to have less prose, more abstraction, and the explicit note that `len(empty_bytes) = 0`.

real-or-random replied:
Ok

real-or-random reviewed (approved):
ACK mod nits :)

robot-dreams commented on bip-musig2.mediawiki:
```diff
@@ -411,7 +421,7 @@ Input:
 * The tweaks ''tweak<sub>1..v</sub>'': ''v'' 32-byte arrays
 * The tweak modes ''is_xonly_t<sub>1..v</sub>'' : ''v'' booleans
-* The message ''m'': a 32-byte array
+* The message ''m'': a byte array with ''0 &le; len(extra_in) &le; 2<sup>32</sup>-1''
```
```suggestion
* The message ''m'': a byte array with ''0 &le; len(m) &le; 2<sup>32</sup>-1''
```

jonasnick replied:
Fixed

robot-dreams commented on bip-musig2.mediawiki:
```diff
@@ -313,17 +314,26 @@ Input:
+** Let ''m_present = bytes(1, 1)''
+* If the optional argument ''extra_in'' is not present:
+** Let ''extra_in = empty_bytestring''
+* Let ''k<sub>i</sub> = int(hash<sub>MuSig/nonce</sub>(rand || bytes(1, len(aggpk)) || aggpk || m_present || bytes(8, len(m)) || m || bytes(4, len(extra_in)) || extra_in || bytes(1, i - 1))) mod n'' for ''i = 1,2''
```
Nit: Would it be worth adding a footnote explaining why `0 <= len(m) <= 2^32 - 1` but we use 8 bytes for encoding `len(m)`?

jonasnick replied:
Good point, this doesn't make sense. I think it's best if we take the same approach with respect to message length as https://github.com/sipa/bips/pull/221.

jonasnick replied:
Actually in contrast to BIP-340 we need to restrict the message length because we have to serialize it within 8 bytes.

jonasnick replied:
I added a commit that extends the message size and adds a footnote that explains that the SHA256 precondition implicitly propagates to the caller. That looks quite weird. It seems to me that it'd be easier here if we'd restric the size of messages in bip340 to 2^48-1 or something.

robot-dreams commented on bip-musig2.mediawiki:
```diff
@@ -313,17 +314,26 @@ Input:
-* Let ''k<sub>i</sub> = int(hash<sub>MuSig/nonce</sub>(rand || bytes(1, len(aggpk)) || aggpk || bytes(1, len(m)) || m || bytes(4, len(extra_in)) || extra_in || bytes(1, i - 1))) mod n'' for ''i = 1,2''
+* If the optional argument ''aggpk'' is not present:
+** Let ''aggpk = empty_bytestring''
+* If the optional argument ''m'' is not present:
```
You want `bytes(1, 0) || bytes(8, 0) || empty_bytestring` in the case where the optional argument `m` was not provided, as opposed to, say, just `bytes(1, 0)`?

I'm ok with this if you are, I just want to confirm.

jonasnick replied:
This seemed to result in the simplest spec. But perhaps not if we replace `m_present` entirely. For example, we could do:
```
    * If the optional argument ''m'' is not present:
    ** Let ''m_prefixed = bytes(1, 0)''
    * Else:
    ** Let ''m_prefixed = bytes(1, 1) || || bytes(8, len(m)) || m''
```
And then only hash `m_prefixed`. I have a slight preference for this.

real-or-random replied:
Suggestion LGTM

robot-dreams reviewed (commented):
Also ACK mod nits

robot-dreams commented:
ACK a668fe122b9b27a517e19d9a4830bdc0726956d3

robot-dreams commented on bip-musig2.mediawiki:
```diff
@@ -313,17 +314,25 @@ Input:
-* The message ''m'': a 32-byte array or 0-byte array (optional argument)
+* The secret signing key ''sk'': a 32-byte array (optional argument)
+* The aggregate public key ''aggpk'': a 32-byte array (optional argument)
+* The message ''m'': a byte array with ''0 &le; len(m) &le; 2<sup>64</sup>-1'' (optional argument) <ref name="mlen">In theory, the allowed message size is smaller because SHA256 accepts byte strings only up to size of 2^56-1 bytes</ref>
```
Footnote copied from https://github.com/sipa/bips/pull/221/

```suggestion
* The message ''m'': a byte array (optional argument) <ref name="mlen">In theory, the message size is restricted due to the fact that SHA256 accepts byte strings only up to size of 2^61-1 bytes.</ref>
```

robot-dreams replied:
Maybe it should also say "restricted due to the 8-byte length encoding"?

jonasnick replied:
Right, good to document that as well.

robot-dreams commented on bip-musig2.mediawiki:
```diff
@@ -352,7 +361,7 @@ The Session Context is a data structure consisting of the following elements:
 * The tweaks ''tweak<sub>1..v</sub>'': ''v'' 32-byte arrays
 * The tweak modes ''is_xonly_t<sub>1..v</sub>'' : ''v'' booleans
-* The message ''m'': a 32-byte array
+* The message ''m'': a byte array with ''0 &le; len(m) &le; 2<sup>64</sup>-1''<ref name="mlen" />
```
```suggestion
* The message ''m'': a byte array<ref name="mlen" />
```

robot-dreams commented on bip-musig2.mediawiki:
```diff
@@ -411,7 +420,7 @@ Input:
 * The tweaks ''tweak<sub>1..v</sub>'': ''v'' 32-byte arrays
 * The tweak modes ''is_xonly_t<sub>1..v</sub>'' : ''v'' booleans
-* The message ''m'': a 32-byte array
+* The message ''m'': a byte array with ''0 &le; len(m) &le; 2<sup>64</sup>-1''<ref name="mlen" />
```
```suggestion
* The message ''m'': a byte array<ref name="mlen" />
```

jonasnick commented:
I removed the explicit restriction for messages and kept the footnote.

robot-dreams commented:
ACK 60b806e18467759cf3f41e3ff94381751fe2eda5
